### ISSUE TITLE
Fatal signal 11 during Android replication

### ISSUE THREAD

**Issue by wj15github** [couchbase/couchbase-lite-android#741]
Hi guys, I use couchbase-lite-android in my Android App to replicate data with my CouchDB server. but I always encounter this problem during replication, here is the error log:
A/libc(10930): invalid address or address of corrupt block 0x79eb9860 passed to dlfree
A/libc(10930): Fatal signal 11 (SIGSEGV) at 0xdeadbaad (code=1), thread 11087 (CBLRequestWorke)

the couchbase-lite-android version is 1.1.0, I implement replication following the official document:
http://developer.couchbase.com/documentation/mobile/1.1.0/develop/guides/couchbase-lite/native-api/replication/index.html

furthermore, there are only 70 documents on my CouchDB server, but couchbase-lite pull replicator shows me there are 592 changes when I invoke getChangesCount(), after I compact and clean up CouchDB server, it also shows that number, what does the change count mean?

any help is appreciated, thanks:)

**Comment by hideki** [comment]
Hi @wj15github 

> furthermore, there are only 70 documents on my CouchDB server, but couchbase-lite pull replicator shows me there are 592 changes when I invoke getChangesCount(), after I compact and clean up CouchDB server, it also shows that number, what does the change count mean?

Couchbase Lite and CouchDB maintain revision history of document. Each document should have one or more revisions. 592 changes indicates total numbers of revisions for 70 documents.

> Hi guys, I use couchbase-lite-android in my Android App to replicate data with my CouchDB server. but I always encounter this problem during replication, here is the error log:
A/libc(10930): invalid address or address of corrupt block 0x79eb9860 passed to dlfree
A/libc(10930): Fatal signal 11 (SIGSEGV) at 0xdeadbaad (code=1), thread 11087 (CBLRequestWorke)

Could you please provide us detail information?
- What version of CouchDB are you ursing? 1.6?
- Android version / Device information.
- Document size, with/without attachment?
- etc

Thanks!

**Comment by Steven-Mark-Ford** [comment]
@hideki 
cc:@wj15github 

I'm also getting Sig 11. This occurs  after my sync runs for about 15 minutes or so (7000+ changes being pulled).

More Info:
-Android 5.1 (Nexus 5)
-CB 1.1.2

Can a document which is too large cause this? or too many documents cause this?
If so what would be the quickest way to check total doc size via rest api? Also I don't know the CouchDB version offhand is there a way to query this via the rest api? Else, where could I get this info?

**Comment by Steven-Mark-Ford** [comment]
@wj15github what device/emulator were you using?

**Comment by hideki** [comment]
Hi @Steven-Mark-Ford,

Thank you for your additional information.

I think you can obtain CouchDB version by REST API

Example
```
curl http://127.0.0.1:5984/
```
```json
{"couchdb":"Welcome","version":"0.10.1"}
```

From crash log, we could know only it crashes during replication from thread id. So we'd like to reproduce the problem. Can you provide us sample documents, number of documents, and average number of changes per documents?


-CB 1.1.2 => we don't have `1.1.2`, Is it `1.1.0`??

Thanks !

**Comment by wj15github** [comment]
Hi @hideki @Steven-Mark-Ford sorry for this late reply

Here is more detail infomation:
- CouchDB v1.6.1
- Android 4.4.2 (Moto G)
- Total database size is 252.1 KB without attachment
- the sig 11 error does not occur every time, it is about 40% percent of tests will get the error. What I do in the test is creating a new Couchbase lite db and replicating with CouchDB. So the App db is empty and most work of replication is just pulling data from CouchDB, right?

**Comment by wj15github** [comment]
by the way, is there anyway that I can only replicate the latest version of documents? Thanks:)

**Comment by hideki** [comment]
@wj15github,

To sync document correctly, it requires document revisions. 

You could post this question on https://forums.couchbase.com . Other engineers might know the trick.

**Comment by wj15github** [comment]
>To sync document correctly, it requires document revisions.
>You could post this question on https://forums.couchbase.com . Other engineers might know the trick.
ok, I will have a try, thanks:)

**Comment by Steven-Mark-Ford** [comment]
We tried on a couple of devices and it only gave us issue on Nexus 5 (as per device spec above)

**Comment by hideki** [comment]
Hi @wj15github,
The feature (https://github.com/couchbase/couchbase-lite-java-core/issues/939) is available from v1.2.0 with Sync Gateway 1.2.0. It might satisfy your requirements.
Thanks!

**Comment by hideki** [comment]
Hi @Steven-Mark-Ford,

v1.2.0 release might solve the problem, please try  http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.2.0-26/ ?

You need to add custom maven repo to your project
See: https://github.com/couchbase/couchbase-lite-android#using-master-branch-version-bleeding-edge

**Comment by zgramana** [comment]
@steve-mark-ford will close this issue out in a week  unless you have an update.

**Comment by zgramana** [comment]
Please refile if the issue occurs again.

**Comment by kostyasev** [comment]
Hi @hideki, not sure is it the same issue but I have got 
A/libc: Fatal signal 6 (SIGABRT) at 0x000024eb (code=-6), thread 9676 (nc-pull-22482-3)
A/libc: Send stop signal to pid:9451 in void debuggerd_signal_handler(int, siginfo_t*, void*)
in my application 
I use com.couchbase.lite:couchbase-lite-android:1.2.1 
and I use continuous replication, 
It happens when the application start and makes the first sync.
could you help me solve this issue?

**Comment by hideki** [comment]
HI @kostyasev,

We need more information what causes SIGABRT.
Do you use forestdb as storage? Do you have more log message before SIGABRT?

1.3.0-12 is Developer preview for 1.3.0 release. Can you use 1.3.0-12 to check if crash still occurs with it?
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.3.0-12/

Thanks!

**Comment by kostyasev** [comment]
@hideki 
thanks, I will try 1.3.0.
"Do you use forestdb as storage?" - did you mean in the server side? 

[fatal signal.txt](https://github.com/couchbase/couchbase-lite-android/files/357619/fatal.signal.txt)
I have attached some log file.
By the way when I used com.couchbase.lite:couchbase-lite-android:1.2.0 I had error like 
 A/libc: Fatal signal 6 (SIGABRT) at 0x0000463f (code=-6), thread 18176 (CBLRequestWorke)
A/libc: Send stop signal to pid:17983 in void debuggerd_signal_handler(int, siginfo_t*, void*)

**Comment by hideki** [comment]
Thank you for additional information.

> "Do you use forestdb as storage?" - did you mean in the server side?

Client side (Couchbase Lite)

**Comment by kostyasev** [comment]
Hello @hideki 
I have tried 1.3.0 release, I have made a couple of tests and the error does not appear.
I notice that some changes was added in logic of the "Replication.ChangeEvent", I mean in 1.2.0 release if the couchbase lite syncs some data then the status often was REPLICATION_ACTIVE
and only after sync completed it was REPLICATION_IDLE. in 1.3.0 release even if couchbase are syncing the data, the status is REPLICATION_IDLE. And it is slightly confusing me.
Thanks for help.
